docs/architecture/adr-024-coin-metadata.md

```diff
@@ -0,0 +1,145 @@
+# ADR 024: Coin Metadata
+
+## Changelog
+
+- 05/19/2020: Initial draft
+
+## Status
+
+Proposed
+
+## Context
+
+Assets in the Cosmos SDK are represented via a `Coins` type that consists of an `amount` and a `denom`,
+where the `amount` can be any arbitrarily large or small value. In addition, the Cosmos SDK uses an
+account-based model where there are two types of primary accounts -- basic accounts and module accounts.
+All account types have a set of balances that are composed of `Coins`. The `x/bank` module keeps
+track of all balances for all accounts and also keeps track of the total supply of balances in an
+application.
+
+With regards to a balance `amount`, the Cosmos SDK assumes a static and fixed unit of denomination,
+regardless of the denomination itself. In other words, clients and apps built atop a Cosmos-SDK-based
+chain may choose to define and use arbitrary units of denomination to provide a richer UX, however, by
+the time a tx or operation reaches the Cosmos SDK state machine, the `amount` is treated as a single
+unit. For example, for the Cosmos Hub (Gaia), clients assume 1 ATOM = 10^6 uatom, and so all txs and
+operations in the Cosmos SDK work off of units of 10^6.
+
+This clearly provides a poor and limited UX especially as interoperability of networks increases and
+as a result the total amount of asset types increases. We propose to have `x/bank` additionally keep
+track of metadata per `denom` in order to help clients, wallet providers, and explorers improve their
+UX and remove the requirement for making any assumptions on the unit of denomination.
+
+## Decision
+
+The `x/bank` module will be updated to store and index metadata by `denom`, specifically the "base" or
+smallest unit -- the unit the Cosmos SDK state-machine works with. The metadata should ideally be
+extensible and serve the needs of many external consumers and even other modules (e.g. `x/ibc` or `x/nft`).
+At a minimum, the metadata type should include all units and a potential list of denomination aliases.
+
+To provide a means of extensibility, metadata may include a non-zero length list of `extensions`,
+which are simply key/value pairs. In addition, metadata may also include a non-zero length list of
+`aliases` with regard to the origin chain (e.g. `["ETH", "ether"]`). Both `extensions` and `aliases`
```

Comment:
> Both `extensions` and `aliases` may be changed via governance.

How would this work?

Reply:
Via a custom governance proposal. One that bank would implement  (e.g. `DenomMetadataProposal`). I intentionally left this out of the scope of the ADR.

Reply:
Aliases may be helpful to provide a convenient name for IBC voucher-tokens tied to another chain.

Reply:
I am not sure metadata should necessarily be created by governance, maybe it can be created by the issuer of a token?

Reply:
Yeah not totally clear on that either. Aliases should not really change -- maybe extensions are modified.